dfdaemon/handler/root_handler.go

```diff
@@ -17,79 +17,104 @@
+
+	hostIP := util.ExtractHost(r.URL.Host)
+	if reg, err = matchRegistry(hostIP, global.Properties.Registries); err != nil {
+		sendResponse(w, http.StatusForbidden, err.Error())
+		logrus.Warnf("%v", err)
+		return
 	}
-	log.Debugf("pre access:%s", r.URL.String())
+	target = proxyURL(r.URL, reg)
 
-	targetURL := new(url.URL)
-	*targetURL = *r.URL
-	targetURL.Path = ""
-	targetURL.RawQuery = ""
+	logrus.Debugf("post access:%s", target)
 
-	hostIP := util.ExtractHost(r.URL.Host)
-	trustHost := global.CommandLine.TrustHosts[hostIP]
-	switch hostIP {
-	case "127.0.0.1", "localhost", global.CommandLine.HostIP:
-		if len(global.CommandLine.Registry) > 0 {
-			targetURL.Host = global.RegDomain
-			targetURL.Scheme = global.RegProto
-		} else {
-			log.Warnf("registry not config but url host is %s", hostIP)
+	reverseProxy := httputil.NewSingleHostReverseProxy(target)
+	reverseProxy.Transport = NewDFRoundTripper()
+	reverseProxy.ServeHTTP(w, r)
+}
+
+func matchRegistry(host string, reg []*config.Registry) (*config.Registry, error) {
+	for i := 0; i < len(reg); i++ {
+		if reg[i].Match(host) {
+			return reg[i], nil
 		}
-	case trustHost:
-		// if the hostIP is trusted, we should forward it directly.
-		targetURL.Host = r.URL.Host
-		targetURL.Scheme = r.URL.Scheme
-	default:
-		// non localhost access should be denied explicitly, otherwise we
-		// are falling into a dead loop: a reverse proxy for itself.
-		// TODO: we do not need such check actually, anything that served
-		// by dfdaemon should only be accessed by localhost which should
-		// be controlled by the listener addr.
-		w.WriteHeader(http.StatusForbidden)
-		log.Warnf("%s is forbidden to forward", hostIP)
-		return
 	}
+	return nil, fmt.Errorf("no matched registry for %s", host)
+}
 
-	log.Debugf("post access:%s", targetURL.String())
+func proxyURL(origin *url.URL, reg *config.Registry) (proxy *url.URL) {
+	if origin == nil || reg == nil {
+		return nil
+	}
 
-	// TODO: do we really need to construct this every time?
-	reverseProxy := httputil.NewSingleHostReverseProxy(targetURL)
+	proxy = new(url.URL)
+	*proxy = *origin
+	proxy.Path = ""
+	proxy.RawQuery = ""
 
-	reverseProxy.Transport = NewDFRoundTripper()
+	if reg.Schema != "" {
+		proxy.Scheme = reg.Schema
+	}
+	if reg.Host != "" {
+		proxy.Host = reg.Host
+	}
+	return proxy
+}
 
-	reverseProxy.ServeHTTP(w, r)
+func amendRequest(r *http.Request) error {
+	if r.URL.Host != "" {
+		return nil
+	}
+	r.URL.Host = r.Host
+	if r.URL.Host == "" {
+		r.URL.Host = r.Header.Get("Host")
+	}
+	if r.URL.Host == "" {
+		// if host is still empty, we need skip to forward it.
+		return fmt.Errorf("url host is empty")
+	}
+	r.Host = r.URL.Host
+	r.Header.Set("Host", r.Host)
+
+	if r.URL.Scheme != "" {
+		return nil
+	}
+	if global.UseHTTPS {
+		r.URL.Scheme = "https"
+	} else {
+		r.URL.Scheme = "http"
+	}
+	return nil
+}
+
+func sendResponse(w http.ResponseWriter, code int, resp string) {
+	w.WriteHeader(http.StatusBadRequest)
```

Comment:
The params `code` of function `sendResponse` will always be re-assigned to `http.StatusBadRequest`?

Reply:
done